go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -128,40 +131,53 @@ func NewNetAttachDefinitionController(name string, ncm NetworkControllerManager,
 	return nadController
 }
 
-func (nadController *NetAttachDefinitionController) Run(stopChan <-chan struct{}) error {
-	defer utilruntime.HandleCrash()
-	defer func() {
-		klog.Infof("Shutting down controller %s", nadController.name)
-		nadController.queue.ShutDown()
-	}()
+func (nadController *NetAttachDefinitionController) Start() error {
+	klog.Infof("Starting %s NAD controller", nadController.name)
+	g := errgroup.Group{}
+	// run on a goroutine so that we can be stopped if we happen to
+	// be stuck waiting for cache sync
+	g.Go(nadController.start)
+	return g.Wait()
+}
 
-	nadController.nadFactory.Start(stopChan)
-	klog.Infof("Starting controller %s", nadController.name)
-	if !cache.WaitForNamedCacheSync(nadController.name, stopChan, nadController.netAttachDefSynced) {
-		return fmt.Errorf("error syncing cache")
+func (nadController *NetAttachDefinitionController) start() error {
+	nadController.nadFactory.Start(nadController.stopChan)
+	if !cache.WaitForNamedCacheSync(nadController.name, nadController.stopChan, nadController.netAttachDefSynced) {
+		return fmt.Errorf("stop requested while syncing caches")
 	}
 
 	err := nadController.SyncNetworkControllers()
 	if err != nil {
 		return fmt.Errorf("failed to sync all existing NAD entries: %v", err)
 	}
 
-	klog.Info("Starting workers for controller %s", nadController.name)
-	wg := &sync.WaitGroup{}
+	klog.Info("Starting workers for %s NAD controller", nadController.name)
 	for i := 0; i < numberOfWorkers; i++ {
-		wg.Add(1)
+		nadController.wg.Add(1)
 		go func() {
-			defer wg.Done()
-			wait.Until(nadController.worker, nadController.loopPeriod, stopChan)
+			defer nadController.wg.Done()
+			wait.Until(nadController.worker, nadController.loopPeriod, nadController.stopChan)
 		}()
 	}
-	wg.Wait()
 
-	// wait until we're told to stop
-	<-stopChan
 	return nil
 }
 
+func (nadController *NetAttachDefinitionController) Stop() {
+	klog.Infof("Shutting down %s NAD controller", nadController.name)
+
+	close(nadController.stopChan)
+	nadController.queue.ShutDown()
```

Comment:
self-note: by moving the logic here, we break that infinite dependency which allows workers to stop properly and allows Start to finish.